### DISCUSSION TITLE
SOCK5 proxy support

### DISCUSSION THREAD

**Discussion by ramazangirgin** [rabbitmq/hop#511] [General]
Hi, 
We use the rabbitmq http api managing the exchanges and queues.
We want to access the rabbitMQ api via SOCK5 proxy with rabbitmq http api.

I checked the support for the library and it is visible as possible to configure HTTP proxy with customProxySelector as below; 

```
 @Nonnull
            @Override
            public List<Proxy> select(@Nonnull final URI uri) {
                final Proxy proxy = new Proxy(Proxy.Type.SOCKS, new InetSocketAddress(socksProxyHost, socksProxyPort));
                return List.of(proxy);
            }
```
            
and registering it with 
```
          final var httpLayerFactory = JdkHttpClientHttpLayer.configure()//
                .clientBuilderConsumer(clientBuilder -> clientBuilder.proxy(socksProxySelector))//
                .create();
        clientParameters.httpLayerFactory(httpLayerFactory);
```
but it doesn't work, it doesn't use configured SOCKS proxy.

The reason is RabbitMQ http client library uses HttpClient and
HttpClient is designed primarily for handling HTTP proxies and ignores SOCKS proxies during the proxy selection process.

see implementation from jdk.internal.net.http.HttpRequestImpl#retrieveProxy ( jdk 21 )
```
    /** Retrieves the proxy, from the given ProxySelector, if there is one. */
    private static Proxy retrieveProxy(ProxySelector ps, URI uri) {
        Proxy proxy = null;
        List<Proxy> pl = ps.select(uri);
        if (!pl.isEmpty()) {
            Proxy p = pl.get(0);
            if (p.type() == Proxy.Type.HTTP)
                proxy = p;
        }
        return proxy;
    }

```
some additional explanation/discussion from stackoverflow.
https://stackoverflow.com/questions/70011046/how-to-use-a-socks-proxy-with-java-jdk-httpclient 

I tried to create a custom HttpLayer to use HttpURLConnection which has SOCKS proxy support instead of  HttpClient but it is not possible to implement custom com.rabbitmq.http.client.HttpLayer. 
The reason is ;
com.rabbitmq.http.client.ParameterizedTypeReference class in the com.rabbitmq.http.client package is not public, which prevents me having a custom HttpLayer implementation.

My question is;
is there way have SOCKS proxy support by using rabbitmq http api ?
Thanks in advance,

Best regards
Ramazan Girgin

**Comment by michaelklishin** [comment]
@ramazangirgin you are welcome to contribute the changes necessary to allow you to implement your own `HttpLayer` with SOCKS5 support.

**Comment by acogoluegnes** [comment]
I made `ParameterizedTypeReference` public, you can implement your own `HttpLayer` now.
